ESA Hubble cone search improvements

jespinosaar opened the pull request:
Cone search now saves the results in a file and a cone_search_criteria method implemented as a joint method for cone search and query criteria searches.

bsipocz commented:
@jespinosaar - this will need a rebase, unfortunately there are a lot of unrelated commits included. 

I'll try to setup a call, or maybe even a meetup during ADASS where we can iterate on a workflow that could work both for the ESAC team and us. Ultimately, I think the vast majority of these rebases could be avoided.

bsipocz commented:
Also, please ping us when this is ready for review.

jespinosaar commented:
> 
> 
> @jespinosaar - this will need a rebase, unfortunately there are a lot of unrelated commits included.
> 
> I'll try to setup a call, or maybe even a meetup during ADASS where we can iterate on a workflow that could work both for the ESAC team and us. Ultimately, I think the vast majority of these rebases could be avoided.

Hi @bsipocz , thanks for your input, it seems that some changes came when we rebased our master fork... Anyway, I will rebase this. And I think that it would be a good idea to discuss about a workflow that could work for both teams, I will tell other Archives about it, thanks!

jespinosaar commented:
Ok @bsipocz , It is now rebased. Let's see if everything is working as expected...

bsipocz commented:
The last merge commit is still superfluous, but let's deal with it only after a round of review.

jespinosaar commented:
> 
> 
> The commits are still around, something must have gone wrong with the rebase. Did you force push the branch back?

Hi @bsipocz, yes, I did, I don't know why they still appear... I have dropped the previous commits (the ones not related to this pull request) but they are still there...

jespinosaar commented:
Hi @bsipocz , this is ready for review, we will see what do to with the rebase.

jespinosaar commented:
Hi again @bsipocz, would you review this before rebasing it?

jespinosaar commented:
It seems it has failed due to some issues related to the CI environment, @bsipocz could you please check it?

keflavich commented:
@jespinosaar We still need to clean up this commit history.  The first 10 commits or so are messy merges that we do not want in our commit history.  I think otherwise we may be ready to merge it

keflavich commented:
Once the commit history is squashed, we'll review this again and likely merge it.

jespinosaar commented:
Sorry for the late response, I have been involved in other tasks. Ok, I have tried to rebase again avoding these previous commits but it seems that now it is taking more changes... could you please help me in this? Thanks!

pep8speaks commented:
Hello @jespinosaar! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`astroquery/esa/hubble/tests/test_esa_hubble.py`](https://github.com/astropy/astroquery/blob/fdb6e1af8089d983b176ee22a17d98fd60622cb6/astroquery/esa/hubble/tests/test_esa_hubble.py):

> [Line 447:68](https://github.com/astropy/astroquery/blob/fdb6e1af8089d983b176ee22a17d98fd60622cb6/astroquery/esa/hubble/tests/test_esa_hubble.py#L447): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

bsipocz commented:
I've done a rebase and pushed it back. This should now both pick up CI and have a clean commit history.

(commands used: `git fetch --all`, `git rebase -i astropy/master` then I deleted all but one set of your commits, saved the file and exited. The rebase was a clean one, no conflicts. Then `git push esdc-esac-esa-int HEAD:ehst -f`

bsipocz reviewed (commented):
Minor comments only, it would be great to have remote_data tests, too, so to make sure this all still works when accessing the archive, too.

jespinosaar commented:
Thanks @bsipocz ! I will include your modifications and I think it is a great idea to include the remote data tests, so I will also upload them in this PR.

jespinosaar commented on astroquery/esa/hubble/core.py:
```diff
@@ -54,10 +57,11 @@ def __init__(self, tap_handler=None):
-    def download_product(self, observation_id, calibration_level="RAW",
-                         filename=None, verbose=False):
+    def download_product(self, observation_id, calibration_level=None,
+                         product_type=None, filename=None, verbose=False):
```
I have used kwarg arguments to solve this

jespinosaar commented on astroquery/esa/hubble/core.py:
```diff
@@ -185,54 +217,197 @@ def get_postcard(self, observation_id, calibration_level="RAW",
-                                 params=payload,
-                                 cache=cache,
-                                 timeout=self.TIMEOUT)
+        radius_in_grades = float(radius/60)  # Converts to degrees
```
I have used astropy coordinates and the set of functions to improve this, thanks!

jespinosaar commented:
I have also added remote_data tests, but as they are not processed in CI, codecov fails in the coverage. Anyway, I have pushed a more simple simple version of remote_data tests and I will improve them in the next pull request (I will ask you how to avoid this issue) , as I will implement a new functionality soon. Is everything ok @bsipocz  ?

bsipocz commented on docs/esa/hubble.rst:
```diff
@@ -239,11 +260,55 @@ If no criteria are specified to limit the selection, this function will retrieve
 
 -------------------------------
-7. Getting access to catalogues
+8. Getting access to catalogues
```
you may consider using rst/sphinx to auto numerate these headings, but it's also fine to leave as is.

jespinosaar replied:
I will leave it as is for this PR, maybe I will remove it in the following iteration (it seems that enumerations are not widely used in other modules...).

bsipocz reviewed (changes requested):
Some minor comments, and one preference to use kwarg only arguments rather than a ``**kwarg`` to store all new arguments. We started to switch over to have them in some modules and I think it's much more cleaner and unambiguous for the users to see the full function signatures rather than letting them pass an arbitrary dictionary.

jespinosaar commented on astroquery/esa/hubble/core.py:
```diff
@@ -185,54 +220,195 @@ def get_postcard(self, observation_id, calibration_level="RAW",
         shutil.move(response, filename)
 
     def cone_search(self, coordinates, radius=0.0, filename=None,
-                    output_format='votable', cache=True):
+                    output_format='votable', async_job=False,
+                    cache=True, verbose=False):
```
Now radius is positional.

jespinosaar commented on astroquery/esa/hubble/core.py:
```diff
@@ -185,54 +220,195 @@ def get_postcard(self, observation_id, calibration_level="RAW",
-            filename = "cone." + str(output_format)
+    def cone_search_criteria(self, target=None, coordinates=None,
+                             radius=0.0,
+                             calibration_level=None,
```
Now radius is positional in cone_search_criteria.

jespinosaar commented on astroquery/esa/hubble/tests/test_esa_hubble.py:
```diff
@@ -48,7 +48,26 @@ def ehst_request(request):
+def ehst_cone_search(request):
+    try:
+        mp = request.getfixturevalue("monkeypatch")
+    except AttributeError:  # pytest < 3
```
Ok, removed!

jespinosaar commented:
Ok, I think I have answered all requests, if I am missing something or if I did not understand some of them, please let me know. Thanks @bsipocz and nice weekend!

jespinosaar commented:
Hi @bsipocz ! I have just uploaded a simple change to include a client_id, but it seems it is failing in vo_conesearch, could you please help me on this? maybe should I rebase this branch again? Thanks for your support!

jespinosaar commented:
Hi @bsipocz, @adamginsburg  ,I saw you released v0.4.2 (congratulations!) so I have rebased the code and modified the changes.rst file accordingly, we would like to close this PR asap, could you please review it and so the proposed changes? Thanks in advance!

keflavich commented on astroquery/esa/hubble/core.py:
```diff
@@ -179,60 +217,202 @@ def get_postcard(self, observation_id, calibration_level="RAW",
+
+        Parameters
+        ----------
+        radius : float
+            radius in arcmin of the cone_search
```
This should read 'float or quantity' - if you give a value in degrees, it will be handled correctly be the code below.

Also, `coordinates` is before `radius` in the arguments, so they should be swapped

jespinosaar replied:
Ok, modified, now we can use a number (in arcmin) or a quantity (in the desired units). Coordinates and radius have also been swapped.

` Parameters
        ----------
        coordinates : astropy.coordinate, mandatory
            coordinates of the center in the cone search
        radius : float or quantity
            radius in arcmin (int, float) or quantity of the cone_search`
            
`        if type(radius) == int or type(radius) == float:
            radius_in_grades = Angle(radius, units.arcmin).deg
        else:
            radius_in_grades = radius.to(units.deg).value`

keflavich replied:
@jespinosaar that change wasn't necessary, `Angle(radius, units.arcmin)` will work if you pass radius as a float or as a quantity:
```python
In [6]: coordinates.Angle(5, unit=u.arcmin)
Out[6]: <Angle 5. arcmin>

In [7]: coordinates.Angle(1/12*u.deg, unit=u.arcmin)
Out[7]: <Angle 5. arcmin>
```

keflavich replied:
I meant for you to only change the documentation, since the code you had before already worked

jespinosaar replied:
Sorry for the misunderstanding, well, now it will distinguish between cases. Maybe I can change it in the next PR. Thanks!

keflavich commented on astroquery/esa/hubble/core.py:
```diff
@@ -179,60 +217,202 @@ def get_postcard(self, observation_id, calibration_level="RAW",
+        if target:
+            try:
+                ra, dec = self._query_tap_target(target)
+            except Exception:
```
does `_query_tap_target` raise a particular exception type?  Please catch only specific exception types; the generic `Exception` here will catch even `ctrl-c` from the user, which we shouldn't do.

jespinosaar replied:
I have modified _query_tap_target to detect a key error when no result is available and raise a ValueError informing that the target cannot be resolved.

keflavich commented on docs/esa/hubble.rst:
```diff
@@ -239,11 +260,63 @@ If no criteria are specified to limit the selection, this function will retrieve
+  >>> from astroquery.esa.hubble import ESAHubble
+  >>> esahubble = ESAHubble()
+  >>> c = coordinates.SkyCoord("00h42m44.51s +41d16m08.45s", frame='icrs')
+  >>> result = esahubble.cone_search_criteria(coordinates=c,radius=7,
```
it would be nice to use `7*u.arcmin` here rather than a raw number

jespinosaar replied:
Modified to include it, based on the first comment.

keflavich reviewed (approved):
I have a few minor comments that could be cleaned up to improve the documentation a bit, but otherwise I think this is good to merge.

jespinosaar commented:
Thanks for your comments and inputs @keflavich! I have integrated the proposed changes and it seems everything is working.

ceb8 commented:
Thanks @jespinosaar!

jespinosaar commented:
Thanks a lot @ceb8 ! See you on next PR!
